# Mass and isotopes

This tutorial includes following fundamental operations related to molecular mass and isotopes.

- Standard molecular weight
- Relative molecular mass (Calculated exact mass)
- Monoisotopic mass
- Isotopic composition
- Simulate mass spectrum

## Loading packages

This tutorial uses `Plots.jl` to display simulated mass spectrum.

In [1]:
using Pkg
Pkg.activate("..")
using MolecularGraph
using Plots

# Convenient function to display a pair of mol images
function displayimgpair(img1, img2)
    """<div>
        <div style="float:left">$(img1)</div>
        <div style="float:left">$(img2)</div>
    </div>"""
end

nothing

 Activating environment at `~/Repository/MolecularGraph.jl_notebook/Project.toml`


## Molecular weight (Relative molecular mass)

Molecular weight (Relative molecular mass) can be calculated by `standardweight` method.

In [2]:
standardweight(smilestomol("CCO"))

(46.069, 0.004529999999995926)

- `standardweight` returns the tuple of standard weight and its uncertainty, respectively. The result shown above means the molecular weight of C2H5O is 46.069±0.005.
- If you dont need uncertainty, give Float64 type to the method. Then the value will be rounded at 2 digits after the decimal point.

In [3]:
standardweight(Float64, smilestomol("CCO"))

46.07

## Calculated exact mass and monoisotopic mass

Both `exactmass` and `monoisotopicmass` returns exact mass of the given molecule. The difference is that `exactmass` considers atomic mass specified in `Atom.mass` fields of atoms whereas `monoisotopicmass` always returns exact mass of the molecule consists of the most abundant isotopes in terrestrial sources.

In [4]:
etohd6 = smilestomol("[2H]C([2H])([2H])C([2H])([2H])O[2H]")

println("Molecular weight: ", standardweight(Float64, etohd6))
println("Monoisotopic: ", monoisotopicmass(Float64, etohd6))
println("Exact: ", exactmass(Float64, etohd6))

Molecular weight: 52.11
Monoisotopic: 46.041865
Exact: 52.079525


## Isotopic composition

`isotopiccomposition(atomsymbol, num)` returns total masses of the given number of atoms and their isotopic compositions.

Isotopes that have lower abundance than the `threshold` parameter will be filtered out (default 0.001 = 0.1%)

In [5]:
isotopiccomposition(:C, 100; threshold=0.01)

5-element Array{Tuple{Float64,Float64},1}:
 (1204.01341934028, 0.01829972319797687)
 (1203.01006450521, 0.06977096630391633)
 (1202.00670967014, 0.19747098047720194)
 (1201.00335483507, 0.36875136628763155)
 (1200.0, 0.3410365354703909)

In [6]:
isotopiccomposition(:C, 1000; threshold=0.01)

15-element Array{Tuple{Float64,Float64},1}:
 (12018.06038703126, 0.011690519214538609)
 (12017.05703219619, 0.019792334300846457)
 (12016.05367736112, 0.0316151371989973)
 (12015.05032252605, 0.04748134937359465)
 (12014.04696769098, 0.06678541451097929)
 (12013.04361285591, 0.08758641222522137)
 (12012.04025802084, 0.1065534144511583)
 (12011.03690318577, 0.11953540283428146)
 (12010.0335483507, 0.12279996975578453)
 (12009.03019351563, 0.11456944727429401)
 (12008.02683868056, 0.09610452890364467)
 (12007.02348384549, 0.0715861101895721)
 (12006.02012901042, 0.04661058275253684)
 (12005.01677417535, 0.02598703123363034)
 (12004.01341934028, 0.012061765298138035)

`isotopiccomposition(mol)` returns isotopic composition of the molecule.

In [7]:
isotopiccomposition(smilestomol("CCO"))

3-element Array{Tuple{Float64,Float64},1}:
 (46.041864812949996, 0.9756627354527866)
 (47.045219648019994, 0.02109374866472255)
 (48.04610980624, 0.0020049807108054704)

## Simulate mass spectrum

According to the isotopic composition, simulated mass spectrum can be plotted by using `simulatemassspec` method and available plot libraries.
Follwing example uses Plots.jl to draw the simulated spectrum.

In [8]:
mol = smilestomol("c1cc(ccc1[C@H]([C@@H](CO)NC(=O)C(Cl)Cl)O)[N+](=O)[O-]")
structure = drawsvg(mol, 200, 200)
data = simulatemassspec(mol)
p = plot(data[:, 1], data[:, 2], size=(320, 240), leg=false, xlabel = "Mass", ylabel = "Intensity")
buf = IOBuffer()
show(buf, MIME("image/svg+xml"), p)
spectrum = String(take!(buf))
close(buf)
display("text/html", displayimgpair(structure, spectrum))

H 
 N + 
 HO 
 Cl 
 O 
 H 
 HO 
 Cl 
 O – 
 NH 
 O 
 
 
 <?xml version="1.0" encoding="utf-8"?>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<path clip-path="url(#clip190)" d="M 0 0 M363.655 808.558 Q367.011 809.275 368.886 811.544 Q370.784 813.812 370.784 817.146 Q370.784 822.261 367.266 825.062 Q363.747 827.863 357.266 827.863 Q355.09 827.863 352.775 827.423 Q350.483 827.007 348.03 826.15 L348.03 821.636 Q349.974 822.771 352.289 823.349 Q354.604 823.928 357.127 823.928 Q361.525 823.928 363.817 822.192 Q366.131 820.456 366.131 817.146 Q366.131 814.09 363.979 812.377 Q361.849 810.641 358.03 810.641 L354.002 810.641 L354.002 806.798 L358.215 806.798 Q361.664 806.798 363.492 805.433 Q365.321 804.044 365.321 801.451 Q365.321 798.789 363.423 797.377 Q361.548 795.942 358.03 795.942 Q356.108 795.942 353.909 796.359 Q351.71 796.775 349.071 797.655 L349.071 793.488 Q351.733 792.748 354.048 792.377 Q356.386 792.007 358.446 792.007 Q363.77 792.007 366.872 794.437 Q369.974 796.845 369.974 800.965 Q369.974 803.835 368.33 805.826 Q366.687 807.794 363.655 808.558 Z" fill="#000000" fill-rule="evenodd" fill-opacity="1" /><path clip-path="url(#clip190)" d="M 0 0 M379.881 823.257 L396.201 823.257 L396.201 827.192 L374.256 827.192 L374.256 823.257 Q376.918 820.502 381.502 815.872 Q386.108 811.22 387.289 809.877 Q389.534 807.354 390.414 805.618 Q391.316 803.859 391.316 802.169 Q391.316 799.414 389.372 797.678 Q387.451 795.942 384.349 795.942 Q382.15 795.942 379.696 796.706 Q377.266 797.47 374.488 799.021 L374.488 794.299 Q377.312 793.164 379.766 792.586 Q382.219 792.007 384.256 792.007 Q389.627 792.007 392.821 794.692 Q396.015 797.377 396.015 801.868 Q396.015 803.998 395.205 805.919 Q394.418 807.817 392.312 810.41 Q391.733 811.081 388.631 814.298 Q385.529 817.493 379.881 823.257 Z" fill="#000000" fill-rule="evenodd" fill-opacity="1" /><path clip-path="url(#clip190)" d="M 0 0 M405.298 823.257 L421.617 823.257 L421.617 827.192 L399.673 827.192 L399.673 823.257 Q402.335 820.502 406.918 815.872 Q411.525 811.22 412.705 809.877 Q414.951 807.354 415.83 805.618 Q416.733 803.859 416.733 802.169 Q416.733 799.414 414.788 797.678 Q412.867 795.942 409.765 795.942 Q407.566 795.942 405.113 796.706 Q402.682 797.47 399.904 799.021 L399.904 794.299 Q402.728 793.164 405.182 792.586 Q407.636 792.007 409.673 792.007 Q415.043 792.007 418.238 794.692 Q421.432 797.377 421.432 801.868 Q421.432 803.998 420.622 805.919 Q419.835 807.817 417.728 810.41 Q417.15 811.081 414.048 814.298 Q410.946 817.493 405.298 823.257 Z" fill="#000000" fill-rule="evenodd" fill-opacity="1" /><path clip-path="url(#clip190)" d="M 0 0 M635.699 808.558 Q639.055 809.275 640.93 811.544 Q642.829 813.812 642.829 817.146 Q642.829 822.261 639.31 825.062 Q635.792 827.863 629.31 827.863 Q627.134 827.863 624.819 827.423 Q622.528 827.007 620.074 826.15 L620.074 821.636 Q622.019 822.771 624.333 823.349 Q626.648 823.928 629.171 823.928 Q633.569 823.928 635.861 822.192 Q638.176 820.456 638.176 817.146 Q638.176 814.09 636.023 812.377 Q633.893 810.641 630.074 810.641 L626.046 810.641 L626.046 806.798 L630.259 806.798 Q633.708 806.798 635.537 805.433 Q637.366 804.044 637.366 801.451 Q637.366 798.789 635.468 797.377 Q633.593 795.942 630.074 795.942 Q628.153 795.942 625.954 796.359 Q623.755 796.775 621.116 797.655 L621.116 793.488 Q623.778 792.748 626.093 792.377 Q628.431 792.007 630.491 792.007 Q635.815 792.007 638.917 794.437 Q642.018 796.845 642.018 800.965 Q642.018 803.835 640.375 805.826 Q638.731 807.794 635.699 808.558 Z" fill="#000000" fill-rule="evenodd" fill-opacity="1" /><path clip-path="url(#clip190)" d="M 0 0 M651.926 823.257 L668.245 823.257 L668.245 827.192 L646.301 827.192 L646.301 823.257 Q648.963 820.502 653.546 815.872 Q658.153 811.22 659.333 809.877 Q661.578 807.354 662.458 805.618 Q663.361 803.859 663.361 802.169 Q663.361 799.414 661.416 797.678 Q659.495 795.942 656.393 795.942 Q654.194 795.942 651.741 796.706 Q649.31 797.47 646.532 799.021 L646.532 794.299 Q649.356